### DISCUSSION TITLE
RFC: Using Performance panel for Node.js / Deno JavaScript CPU Profiling

### DISCUSSION THREAD

**Discussion by jecfish** [ChromeDevTools/rfcs#2] [RFCs]
**Author**: Nancy Li (@lilysjtu2011), Jecelyn Yeen (@jecfish)
**Posted**: Nov 9, 2022
**Status**: Complete
**Chrome bug tracker**: https://crbug.com/1354548

The goal of this RFC is to validate the design with the community, solicit feedback on open questions, and enable experimentation via a non-production-ready prototype included in this proposal.

**Complete**: This RFC is now complete. See a summary https://github.com/ChromeDevTools/rfcs/discussions/2#discussioncomment-7862319.

## Motivation <a href="#user-content-motivation" id="motivation">#</a>

Since [Chrome 58](https://developer.chrome.com/blog/devtools-javascript-cpu-profile-migration-2/), the DevTools team set a long-term goal to deprecate the JavaScript Profiler (Profiler tab), and migrate Node.js and Deno developers to use Performance panel for profiling JavaScript CPU performance.

These changes allow users to profile performance consistently with one single entry point - the Performance panel. 

## Goals <a href="#user-content-goals" id="goals">#</a>

- **Profile Node.js / Deno JavaScript CPU performance with Performance panel** 
  * Developers are able to profile JavaScript CPU performance with the Performance panel.
  * The Performance panel should display relevant JavaScript CPU information, similar to the Profiler tab.

- **Minimize impact on the ecosystem**
  * Overall mental model: developers should be able to apply their existing knowledge to profile JavaScript CPU performance using the Performance panel, with minimal learning curve.

## Current state <a href="#user-content-current" id="current">#</a>

Here is the current workflow on profiling JavaScript CPU performance:

1. Run `node --inspect file-name.js` or `deno --inspect file-name.js` (e.g. use this [example](https://github.com/jecfish/js-deprecate/blob/main/index.js) file)
2. Open DevTools for Node via one of these options:
    - The green Node button at the top of DevTools
    - Enter url `chrome://inspect` > **Devices** > click **Open dedicated DevTools for Node**
    - Enter url `chrome://inspect` > **Remote target** > click **Inspect**
     ![Open DevTools for Node via one of these options](https://user-images.githubusercontent.com/5917927/201757433-7a2daf14-215e-42ca-aff9-e5e1a7a9fee8.png)
3. Start and end profiling by clicking the `Record` button in the Profiler tab.
    ![The Record button in the Profiler tab](https://user-images.githubusercontent.com/5917927/201758270-5d408641-29de-4f85-b741-69695055da7d.png)
4. View the result in the Profiler tab. Use the dropdown to switch to different views. It support 3 modes currently :
    - **Chart**. Users can switch to the Chart view with the dropdown.
    - **Heavy (bottom up)**. This is the default view.
    - **Tree (top down)**. Users can switch to the Chart view with the dropdown.

    *Chart*
    ![Chart](https://user-images.githubusercontent.com/5917927/201839140-42597d80-abe2-4aa1-9831-43e68bc96bb5.png)
    *Heavy (bottom up)*    
    ![Heavy (bottom up)](https://user-images.githubusercontent.com/5917927/201756529-e657b6d5-c85a-4324-90b3-afaa2bc797a3.png) 
    *Tree (top down)*
    ![Tree (top down)](https://user-images.githubusercontent.com/5917927/201839357-6404d2ad-3668-47f6-b281-ca847dbf9851.png)
 

## Try the prototype: Profile using the Performance panel <a href="#user-content-prototype" id="prototype">#</a>

There is a prototype implementation. Below, we show a guide for trying it out. This is a draft implementation, with missing features and non-final design aspects.

Please use Chrome Canary to play with the prototype.

### Profiling with the Record button

1. Run `node --inspect file-name.js` or `deno --inspect file-name.js`
2. Open DevTools for Node. (Chrome Canary)
3. Open the Performance panel via the **3-dot** menu > **More tools** > **Performance** panel. This step is not required once the changes are rolled out officially, the Performance panel will show by default.
    ![Open Performance panel](https://user-images.githubusercontent.com/5917927/201840029-31d2ce7f-ca48-4618-a144-9ddcb93be2cd.png)
4. Start and end profiling by clicking the `Record` button in the Performance panel.
    ![The Record button in the Performance panel](https://user-images.githubusercontent.com/5917927/201757089-5e984dad-9a09-4ef6-8b7d-dbcacdc3573a.png)
5. View the result in the Performance panel. The UI is different from the Profiler tab but you can find all the information:
    - **Chart**. The Performance panel always shows the **chart** on top.
    - **Heavy (bottom up)**. The Performance panel shows this information in the bottom pane > **Bottom-up**.
    - **Tree (top down)**. The Performance panel shows this information  in the bottom pane > **Call tree**.
    
    *Chart*
    ![Chart](https://user-images.githubusercontent.com/5917927/201841002-05cad6d3-b4b6-49c5-aaa6-cc180ee30594.png)
   *Bottom-up*
    ![Bottom-up](https://user-images.githubusercontent.com/5917927/201840773-3b80e5cb-856a-4811-801f-d09c544c31c4.png)
    *Call tree*
     ![Call tree](https://user-images.githubusercontent.com/5917927/201756502-ff64e52f-966b-44ea-ba75-a349a9188590.png)

### Profiling with the console.profile() command

If you profile performance with the `console.profile()` command, you need to enable an experiment to map the result to the Performance panel. 

1.  Enable the **View console.profile() results in the Performance panel for Node.js** checkbox via the **Settings** icon > **Experiments** tab. 
    ![Enable the experiment](https://user-images.githubusercontent.com/5917927/201841881-9fea4dbb-dd95-4b50-bbe4-d06f67957419.png)  
3. Open DevTools for Node. (Chrome Canary) 
4. Run `node --inspect file-name.js` or `deno --inspect file-name.js` (e.g. use this [example](https://github.com/jecfish/js-deprecate/blob/main/index-profile.js) file)
5. Once the profile is completed, the result will be shown in the Performance panel automatically. 

## FAQ <a href="#user-content-faq" id="faq">#</a>

1. **What will happen with the Profiler tab in the future? Will that be deprecated / removed?**
Yes, we plan to deprecate the Profiler tab by stages. We will look into community feedback, and implement the JavaScript profiling capability into the Performance panel before deprecation.

2. **Can I load the CPU profile I saved in the Profiler tab previously into the Performance panel?**
Yes. You can import `.cpuprofile` files to the Performance panel.

3. **Are both Profiler tab and Performance panel using the same data source?**
Yes. The backend data source is the same for both. However, there are frontend logics that change the displayed data in the Profiler tab.

    Below are the same profiling results shown in the Profiler tab and Performance panel.

    ![Profiler tab](https://user-images.githubusercontent.com/5917927/201843158-1f1da448-4f79-4e9a-8ec1-c2cb5ce96037.png)
    ![Performance panel](https://user-images.githubusercontent.com/5917927/201756486-3eaeea11-9dce-41e4-a034-7998e171490f.png)

    There are ~3~ 1 known differences at the moment:

    a. **From showing frames directly to organizing them under call stack** - In the above example, the Profiler shows `executeUserEntryPoint` and the `Module` frames straight away. On the other hand, the Performance panel organizes them under the same call stack, you need to expand them to view it.
    b. ~[crbug.com/1382733](https://crbug.com/1382733) (Fixed) - Some frames like `(garbage collector)` are hidden now in the Performance panel. We are working on adding them.~
    c. ~[crbug.com/1080918](https://crbug.com/1080918) (Fixed) - Differences in profiling `setTimeout` call due to the stack trace collection. We are working on addressing this.~

4. **Where is the realtime total JS heap size view?**
The realtime Total JS heap size information is already available in the Memory panel. There is no plan to move that into the Performance panel.

    *The realtime total JS heap size in the Memory panel*
        ![JS heap size view in the Memory panel](https://user-images.githubusercontent.com/5917927/203033559-662fdae3-30e8-4e4c-b5e2-837a267e00be.png)

5. **Does the Performance panel support reload and record Node.js / Deno performance?**
No. This proposal focuses on migrating the existing features in the Profiler tab to the Performance panel. The Profiler tab doesn’t support reload and record.

## Questions for Discussion <a href="#user-content-questions" id="questions">#</a>

In addition to general feedback, we would like to collect feedback on the following specific questions:

1. **Does the prototype correctly display JavaScript profiling information?**
We are interested in any cases where the information is missing or has discrepancy with the Profiler tab.

2. **Does profiling in the Performance panel UI work well for you?**
We would be interested in any use cases where this change in UI would be problematic or burdensome, and where the UI is important.
For example, should we enlarge the Bottom-up tab and show it by default when you open the Performance panel in Node.js / Deno mode?

**Comment by hashseed** [comment]
@joyeecheung @mhdawson @mcollina @jasnell @gireeshpunathil @qard FYI

**Reply by Qard** [reply]
Looks very nice!

You mention there's no plan to move Total JS heap size, but the performance tab includes memory information currently. Is there some technical reason for _not_ moving Total JS heap size too or just a scope concern and could therefore happen separately in the future? I would really like to see a more unified view for that. Something more like macOS Instruments would be really nice where you could pick which profilers to activate and it will capture and visualize them all together.

**Reply by lilysjtu2011** [reply]
Hi @Qard , the `total JS heap size` in the RFC referring to the **real time total JS heaps size view** (as highlighted below). 

We don’t plan to migrate that to the Performance panel because the Memory panel has a similar UI for that. 

Are you referring to another “Total JS heap size” or you would like to see this information in the Performance panel?
<img width="887" alt="image7" src="https://user-images.githubusercontent.com/28817075/203174262-45b5ca47-4b7a-49f4-9257-0ad6997a222e.png">

**Reply by Qard** [reply]
Yes, that's what I was referring to. The Performance tab has a checkbox for memory and I'm unclear what, if anything, is stopping the existing memory tab data from being ported into that.

**Comment by jecfish** [comment]
@jrieken @bartlomieju fyi

**Comment by gireeshpunathil** [comment]
@rafaelgss fyi

**Comment by joyeecheung** [comment]
Nice write-up of the new performance panel! I am wondering, is the Node.js-dedicated panel going to remain in `chrome://inspect`?

Also,  I gave it a try and profiled a typical HTTP server:

```js
'use strict';

const http = require('http');
const bodyLength = 12345;

const body = 'c'.repeat(bodyLength);
const server = http.createServer(function(req, res) {
  res.writeHead(200, {
    'Content-Length': bodyLength,
    'Content-Type': 'text/plain'
  });
  res.end(body);
});
server.listen(0, '127.0.0.1', () => {
  const { address, port } = server.address();
  console.log(`http://${address}:${port}`);
});
```

And profiled it with:

```
$ autocannon -d 10 http://127.0.0.1:$PORT
```

Here is the comparison (the workload in the particular span that the profile is taken differed slightly so minor differences in the calculations can be ignored):

The call trees:
<img width="480" alt="Screen Shot 2022-11-17 at 22 44 55" src="https://user-images.githubusercontent.com/4299420/202566467-4805c131-ba19-4566-8425-bb1bab7e48de.png">
<img width="510" alt="Screen Shot 2022-11-17 at 22 42 18" src="https://user-images.githubusercontent.com/4299420/202566402-bb96ae04-719f-452c-815f-d0c443e7840c.png">

The bottom-ups:
<img width="505" alt="Screen Shot 2022-11-17 at 22 42 04" src="https://user-images.githubusercontent.com/4299420/202566406-9f164a56-5f72-4961-9805-e5d638939720.png">
<img width="521" alt="Screen Shot 2022-11-17 at 22 41 45" src="https://user-images.githubusercontent.com/4299420/202566713-580edd93-60dc-4695-9055-3eaaa2224090.png">

Close-up looks of some frames:
<img width="891" alt="Screen Shot 2022-11-17 at 22 59 02" src="https://user-images.githubusercontent.com/4299420/202568883-8b914600-e295-4bbb-9081-63bde87dc5e7.png">
<img width="1242" alt="Screen Shot 2022-11-17 at 22 58 38" src="https://user-images.githubusercontent.com/4299420/202568887-9ad7e9b1-7af6-4070-b9e0-c3f6d7650137.png">

Other than the `(garbage collector)` entries missing, as already mentioned in the OP, there are some noticeable differences:

1. `(idle)` is missing
2. `(program)` is missing - like `(idle)` or `(garbage collector)` this is actually in the inspector protocol response. 
3. There is now a `Function call` entry, is there some documentation about what that is? It looks like some sort of "native-calling-back-to-JS trampoline" entry.
4. I can see in the summary that there are `System` and `Idle` entries in the chart, but they are quite different from the calculations in the profiler panel - AFAIK the calculations from the profiler panel are correct (if we replace `System` with `program`, which in V8 means a combination of `EXTERNAL` (embedder) + `PARSER` + `COMPILER` and a few other VM states in the "system").
    <img width="384" alt="Screen Shot 2022-11-17 at 23 25 28" src="https://user-images.githubusercontent.com/4299420/202573315-5d7ba01e-7eaa-4eaa-a721-b6f606a3e70f.png">    
5. It is now possible to see the call trees and bottom-ups of a specific span, which is pretty nice 👍 
6. When I saved the profile from the performance panel, the resulted file is named `Profile-<timestamp>.json`, from the profiler panel it's `CPU-<timestamp>.cpuprofile`. The JSON saved from the performance panel is in a format different from the JSON-formatted response from the inspector protocol (that is, https://chromedevtools.github.io/devtools-protocol/tot/Profiler/#type-Profile). There are some external tools that analyzes the profile using the inspector protocol format, so to fully replace the original panel it would be nice to be able to save the profile in that format, maybe after answering a prompt before saving.

**Reply by lilysjtu2011** [reply]
Hi Joyee, for your first question not sure if I understand your question correctly. Do you mean if we will remove the link in `chrome://inspect`? (See the image below)

The “Open dedicated DevTools for Node” link will remain showing. However, instead of showing the Profiler tab when clicked, it will show the Performance panel instead. 

The Profiler tab will be removed in the future.

<img width="569" alt="image2" src="https://user-images.githubusercontent.com/28817075/203174448-7a6c7494-21ff-4025-b8c7-56b902279526.png">

**Reply by lilysjtu2011** [reply]
For your detailed feedback, here are my replies:

```
Q1 & Q2: Missing (idle)(program) entries
```
These missing entries are caused by the same reason as `(garbage collector)`. We are currently working on it. Here is the tracking bug [crbug.com/1382733](https://crbug.com/1382733).

We will update again when the bug is fixed!

```
Q3: The Function call entry
```
After some investigation, this is actually related to Q1 & Q2 [crbug.com/1382733](https://crbug.com/1382733). After the bug fix, the `Function call` should be gone. 

```
Q4: Incorrect calculation in `System` and `Idle` calculations in the Performance panel’s Summary tab
```
We’ve opened a bug [crbug.com/1392436](https://crbug.com/1392436) for this. I will investigate this further.

```
Q5: Ability to view Call tree and Bottom-ups on a specific span
```
Good to know. Thank you. 

```
Q6: Options to download .cpuprofile from the Performance panel
``` 
Nice feature request. Created a new feature request [crbug.com/1392446](https://crbug.com/1392446) to track this.

**Reply by lilysjtu2011** [reply]
Hey @joyeecheung , some updates here.

The Performance panel is updated so the system nodes like (idle), (garbage collector) and (program) are no longer missing. Also the time calculation is also corrected now.
And for downloading, you can now download `.cpuprofile` when you use the Performance panel for Node.js.

Please take a look at the updates and feedback is welcomed :)
Thanks

**Comment by mcollina** [comment]
A use case I have been using is to collect CPU and memory profiles using the inspector API directly and then loading them up in chrome to visualize them. Given the format has changed, will this still be supported?

**Reply by jrieken** [reply]
Same here. In VS Code we automatically perform CPU profiling and analyse its data. For that we use `Profiler.start` and `Profiler.end` command (from https://chromedevtools.github.io/devtools-protocol/tot/Profiler/). So, is this a UX only change or will the Profiler also disappear from the devtools-protocol?

**Reply by hashseed** [reply]
CDP is not going to be affected. We are talking about replacing one consumer of the protocol with another consumer, for the Node.js use case.

**Comment by Qard** [comment]
Seems to not handle large profiles as well as the old Profiler tab. I opened a ~25mb profile in the Profiler tab which loaded and displayed fine and could be navigated relatively smoothly with only a little bit of lag. I loaded the same profile in the Performance tab and it seemed to lock up while loading, eventually displaying the thing after several minutes. When I tried to click anything in the chart visualization it crashed the whole browser. 😬

**Reply by jecfish** [reply]
Thanks Stephen, created a [crbug.com/1391853](https://crbug.com/1391853) for @lilysjtu2011 to follow up.

**Reply by Qard** [reply]
Thanks. I added a comment that I can share the profile data with V8 folks if they want to investigate that profile specifically. 😄

**Reply by hashseed** [reply]
It's not actually V8 folks, but DevTools folks working on this.

So if you can share your profile with e.g. jec@google.com, we might use the Performance panel on the Performance panel so that we can profile the performance of the Performance panel loading your performance profile.

😄

**Reply by jecfish** [reply]
Hah, before this message, I've dropped @Qard a follow up email with @lilysjtu2011. :)

**Reply by Qard** [reply]
Sorry, that's what I meant, devtools folks, the V8-adjacent. 😅

In any case, I sent the profile along already. Hopefully that proves helpful in improving the performance. 🙂

**Comment by bartlomieju** [comment]
Hi @jecfish, I'm gonna need one or two more days to provide my feedback - thanks to your example I discovered that Deno's console implementation if missing some console methods that should be available in the inspector.

**Comment by jecfish** [comment]
Hi all,

Happy New Year! Thanks for your testing and feedback. @lilysjtu2011 has made a couple of significant improvements based on your feedback. 

Please give it another round of test in Canary and comment!

Below are the improvements:

- Added options to download .cpuprofile from the Performance panel - [crbug.com/1392446](http://crbug.com/1392446)
- Included missing events (e.g. (program), (idle), (garbage collector)) in the Performance panel - [crbug.com/1382733](http://crbug.com/1382733)
- Fixed deviation in System and Idle calculation - [crbug.com/1392436](http://crbug.com/1392436)
- Fixed inaccurate (amplified) setTimeout calls in performance profiles - [crbug.com/1080918](http://crbug.com/1080918)
- [Reporter mentioned not reproduce-able, Let us know if you face this issue] Not handle large profiles well in the Performance panel - [crbug.com/1391853](http://crbug.com/1391853)

Looking forward to your feedback on this new workflow!

**Comment by jecfish** [comment]
Hi all,

We appreciate your valuable testing and feedback thus far. With all blocking issues addressed, we are currently in the process of deprecating the JS Profiler, which involves 4 stages:

1. In Chrome 112, we will display a warning banner in the JS Profiler. Please share any feedback you have in this RFC. 
    <img width="860" alt="Screenshot 2023-03-03 at 12 49 01 PM" src="https://user-images.githubusercontent.com/5917927/222714206-8243c432-bf9a-4de5-96f1-326f6f11d732.png">
2. In Chrome 113, we will disable the tab. However, you will still be able to turn it on temporarily. To provide feedback, please visit [crbug.com/1354548](http://crbug/1354548).
3. In Chrome 114, the tab will be hidden by default. But you can still turn it on temporarily using a flag. Feedback can be shared via [crbug.com/1354548](http://crbug/1354548).
4. Finally, in Chrome 115 or later, we will completely remove the JS Profiler once all blocking issues have been resolved.

If you come across any issues during this process, please do report them to us.

Regards,
*Nancy & Jecelyn*

**Comment by stoem** [comment]
The old JS Profiler allows fairly easy access to the most time consuming functions and even lines of code. 
The same tasks seems much trickier in the Performance Panel - can you point out how we'd achieve narrowing down any performance bottlenecks back to the file/function/line of code that causes it using the Performance Panel?

**Reply by lilysjtu2011** [reply]
Hi @stoem , you can check here:
![image](https://github.com/ChromeDevTools/rfcs/assets/28817075/cd214127-87cc-479c-82ee-2fee22beaf65)
The `Heavy (Bottom Up)` tab in JS profiler is now `Bottom-up`, and `Tree (Top Down)` is now `Call tree`

**Comment by realyze** [comment]
There's two things I would sorely miss if we deprecate JS profiler:

1. If I understand correctly, the Profiler tab only shows information for the selected task, no? What I _very_ often do is that I record an interaction (e.g. dragging or typing) in the JS Profiler tab and then look at total CPU time spent running a give method. The execution time varies between tasks so it's super useful to be able to see total time spent, not just time spent within selected task.

2. The JS profiler tab is _much_ faster. For a long complex profile, it can take tens of seconds before the Profiler tab lets me select any task without deselecting it immediately (possibly a profiler bug?)

**Reply by lilysjtu2011** [reply]
Hi @realyze ,
For 1, can you give me more information? You can see the total time in the `Bottom-Up` or `Call Tree` in the bottom drawer. Don't know if this is what you are finding.
For 2, we are working on improve the speed. Some track bug is here: crbug.com/1457639

**Comment by o0x2a** [comment]
How to select a VM instance in Performance panel? Currently it is possible to select an iframe or web-worker in Profile tab.

**Reply by lilysjtu2011** [reply]
Hi @o0x2a , I don't know if you are using the Profile tab for web app or Node app. In the Performance panel, we currently will automatically capture JS profiles of all JS isolates, and web workers and Out-of-Process iframes are shown in different tracks. Can you provide more information to me?
Thanks

**Reply by tophf** [reply]
@lilysjtu2011, the point is to allow us to select the isolate (the worker), which is essential when there are many isolates/frames that we don't want to profile.

**Reply by jecfish** [reply]
Open a bug to track the investigation: [crbug.com/1511813](crbug.com/1511813)

**Comment by aspin** [comment]
Are there any types of targets that aren't supported by the Performance panel that are by the older Profiler tab? Currently, the profiler tab works for me, but when I try to record a profile in the Performance panel I get "Performance trace recording not supported by this type of target".

**Reply by lilysjtu2011** [reply]
Hi @aspin ,
In Performance panel we also support tracing. Can you give me more information about how you use the Performance panel, and on which target you tried to record the profile? 
Thanks

**Reply by aspin** [reply]
unfortunately I don't have access to the codebase anymore, but it wasn't anything particular complex. I started a node16 app, enabled the debugger and tried to connect to it with the performance panel, and nothing would work. The old profiler connected fine.

**Reply by lilysjtu2011** [reply]
Thanks for the reply, I will try to repro and see if i can get some more information

**Reply by jecfish** [reply]
Opened a bug for further investigation: https://crbug.com/1511486

**Reply by and-oli** [reply]
Hi @aspin.  As with some of the other threads, we have landed a good amount of fixes to the Performance panel that tackle issues like the one you described so, could we please ask you to confirm if the issue persists today?. The latest improvements are shipping from Chrome version 121.0.6167.16 (currently Chrome Beta), so I suggest testing using this. Thanks!

**Comment by suzmas** [comment]
The Performance tab is significantly slower (as previously noted https://github.com/ChromeDevTools/rfcs/discussions/2#discussioncomment-5632504), consumes a ton of CPU for complex profiles, and yesterday it crashed several times while trying to load the output from a fairly long-running profile - I couldn't even download the profile before the Performance tab became unresponsive, so I had to re-run it and take small (sub 90s) profiles multiple times, for the duration of the task. That is, of course, very inconvenient - re-enabled the Profiler tab because of this.

**Reply by lilysjtu2011** [reply]
Hi @suzmas , thanks for your feedback, we are now working on improve the speed. Some track bug is here: crbug.com/1457639. 
Will update here when we have more progress. Thanks

**Comment by Eric-Gurt** [comment]
JS Profiler works on big projects (games) no matter how long you profile it for.

Performance tab just dies after you stop it past more than 5 seconds of profiling within same project (without screenshots or anything else pretty much).

But interestingly profiling with JS Profiler and then exporting/importing same file via Performance tab does work fine too.

**Reply by lilysjtu2011** [reply]
Hi @Eric-Gurt , thanks for your feedback. Can you give more information how you profile the code, and if possible some code example will be helpful. 
Thanks

**Reply by Eric-Gurt** [reply]
This only happens in a big project which I'm not able to share right now. It includes lots of JS files (loaded dynamically), lots of different (usually sealed) objects, huge arrays, THREE.js, canvas-generated textures, geometry manipulations, DataView-s, lots of strings. There is like 21 MB of just JavaScript files.

I just tried to test it with Performance tab and ~15 seconds of recording has been decoded in approximately 1 minute.

And then after tab close and reopening same scenario with JavaScript Profiler tab only took 2-3 seconds after recording was stopped.

So I pressed Ctrl + I on the DevTools for page and right before I stopped Performance tab recording I started profiling this DevTools for page with "JavaScript Profiler" tab... I've got some useful results perhaps?

This is in Edge:
[screenshot](https://imgur.com/4f6Zbf2)
[.cpuprofile](https://www.gevanni.com/CPU-20230715T000643.cpuprofile)

This is in Chrome. It is just my project isn't fully configured for it. Yet still take 20 seconds to show results in Performance tab when it takes nothing when done with JavaScript Profiler:
[screenshot](https://imgur.com/e0XAJNP)
[.cpuprofile](https://www.gevanni.com/Chrome%20CPU-20230715T002004.cpuprofile)

It seems to be proportional to the amount of stuff done by JavaScript on a page (Chrome's case is loading JS files dynamically, does a few requests then show simple THREE.js scene and that is pretty much all it does).

**Reply by lilysjtu2011** [reply]
Thanks for the reply, I will take a look on this.

**Reply by jecfish** [reply]
Open a bug to track the investigation: https://crbug.com/1511491

**Reply by and-oli** [reply]
Hi @Eric-Gurt. Could we please ask you to confirm if the issue persists today? We have landed a good amount of fixes to the Performance panel that tackle issues like the one you described. The latest speed improvements are shipping from Chrome version 121.0.6167.16 (currently Chrome Beta), so I suggest testing using this. Thanks!

**Comment by ajuni880** [comment]
Hi, I am trying to debug a node application it used to work fine with JS Profiler but with Performance tab it just dies after couple of seconds. Also when starting the profiling in performance tab it stops the task initiated with `node inspect-brk`. Stopping manually the profiling shows this screen and it stays here forever.

Chrome Version 114.0.5735.199 (Official Build) (64-bit)

![image](https://github.com/ChromeDevTools/rfcs/assets/22448652/86fa7484-8507-4e0a-bb38-9952392562b2)

**Reply by lilysjtu2011** [reply]
Hi @ajuni880 , thanks for your feedback. Can you give more information about how the Performance tab dies, is there any message in the dialog? And do you know if you profile the same code with the JS Profiler, how big is the file? Also if possible the some example cpuprofile downloaded from JS Profiler or some code example will be helpful. 
Thanks

**Reply by Silvrbckw** [reply]
To be honest I don't know. I have been teaching myself this github process. I don't know if anything is working orn hindering at this point.

**Reply by ajuni880** [reply]
Hi @lilysjtu2011, there is no other message, the file is not that big, it's a jest file with angular tests. I can't download the cpuprofile because devtools hangs up.

**Reply by lilysjtu2011** [reply]
Thanks for the reply, I will try to check this.

**Reply by jecfish** [reply]
Open a bug report to track the investigation: https://crbug.com/1511489

**Reply by and-oli** [reply]
Hi @ajuni880, sorry for jumping so late to the discussion. We have recently worked on improving the support for node profiling use cases in the Performance Panel. As such, I would expect your use case to work fine now on the latest Chrome Stable version. Could we please ask you to confirm if the issue you describe has been fixed? Thanks!

**Reply by and-oli** [reply]
I forgot to mention: the latest Chrome Stable version I was talking about is 120.0.6099.71 (Official Build). Here the Performance Panel should be able to load CPU profiles. However, the latest speed improvements are shipping from version 121.0.6167.16 (currently on Beta), so I suggest testing using this latter.

**Comment by tophf** [comment]
The old profiler immediately opens the linked source code when you click a block in the flamechart.
It'd be nice to have the same via double-click or Ctrl-click or right-click menu in the new profiler.

Currently we have to separately click the source link in the bottom drawer, which is pretty far from the flamechart, and if we switch the tab in the drawer to investigate the block we'll have to return to the Summary tab again, constantly switching tabs is inconvenient.

**Reply by jecfish** [reply]
Open a bug to track the investigation: [crbug.com/1511814](crbug.com/1511814)

**Comment by jecfish** [comment]
After almost a year of discussion, we're closing the **RFC for using the Performance panel for Node.js/Deno JavaScript CPU profiling**. 

As [previously](https://github.com/ChromeDevTools/rfcs/discussions/2#discussioncomment-5189668) communicated, we're currently in stage 3 of the 4-stage JS Profiler deprecation. In upcoming releases, **we'll remove the JS Profiler entirely once all blocking issues are resolved**.

**Blocking Issues**

- **Performance**. We’ve fixed the major performance issue that is blocking you. From our benchmark, The speed has improved significantly in the Performance panel. For example, the initial loading time for a 48m file, dropped from 10s to 2s. Please give it a spin and let us know! ([crbug.com/1457639](http://crbug.com/1457639))
- **Profile selected VM instance**. We understand that the JS Profiler allows you to select the VM instance for profiling. We're actively working on implementing it within the Performance panel before the final removal of the JS Profiler. ([crbug.com/1511813](http://crbug.com/1511813))

**Follow ups**

There are other [UX improvement suggestions](https://bugs.chromium.org/p/chromium/issues/list?q=blocking%3A1354548&can=2) as well, we’ll consider adding them, possibly in follow-up releases.

If you have any questions or feedback, please share them at [crbug.com/1354548](http://crbug.com/1354548).

We would like to thank everyone for your feedback and for your help in shaping Chrome DevTools!
